File: odemapster-r2rml-rdb/src/main/java/es/upm/fi/dia/oeg/obdi/wrapper/r2rml/rdb/model/R2RMLJoinCondition.java
package es.upm.fi.dia.oeg.obdi.wrapper.r2rml.rdb.model;

import java.util.Collection;

import org.apache.log4j.Logger;

import Zql.ZConstant;
import Zql.ZExpression;

import com.hp.hpl.jena.rdf.model.Resource;
import com.hp.hpl.jena.rdf.model.Statement;

import es.upm.fi.dia.oeg.morph.base.Constants;
import es.upm.fi.dia.oeg.obdi.wrapper.r2rml.rdb.exception.R2RMLJoinConditionException;

public class R2RMLJoinCondition {
	private static Logger logger = Logger.getLogger(R2RMLJoinCondition.class);
	
	private String childColumnName;
	private String parentColumnName;
	
	
	public R2RMLJoinCondition(String childColumnName, String parentColumnName) {
		super();
		this.childColumnName = childColumnName;
		this.parentColumnName = parentColumnName;
	}
	
	public R2RMLJoinCondition(Resource resource) throws R2RMLJoinConditionException {
		Statement childStatement = resource.getProperty(Constants.R2RML_CHILD_PROPERTY());
		if(childStatement != null) {
			this.childColumnName = childStatement.getObject().toString();
		} else {
			String errorMessage = "Missing child column in join condition!";
			logger.error(errorMessage);
			throw new R2RMLJoinConditionException(errorMessage);
		}

		Statement parentStatement = resource.getProperty(Constants.R2RML_PARENT_PROPERTY());
		if(parentStatement != null) {
			this.parentColumnName = parentStatement.getObject().toString();
		} else {
			String errorMessage = "Missing parent column in join condition!";
			logger.error(errorMessage);
			throw new R2RMLJoinConditionException(errorMessage);
		}
		
	}

	@Override
	public String toString() {
		return "JOIN(Parent:" + this.parentColumnName + ",CHILD:" + this.childColumnName + ")";
	}

	public String getChildColumnName() {
		return childColumnName;
	}

	public String getParentColumnName() {
		return parentColumnName;
	}
	
	public static ZExpression generateJoinCondition(
			Collection<R2RMLJoinCondition> joinConditions, String parentTableAlias
			, String joinQueryAlias, String dbType) {
		ZExpression onExpression = null;
		String enclosedCharacter = Constants.getEnclosedCharacter(dbType);
		
		if(joinConditions != null) {
			for(R2RMLJoinCondition joinCondition : joinConditions) {
				String childColumnName = joinCondition.getChildColumnName();
				childColumnName = childColumnName.replaceAll("\"", enclosedCharacter);
				childColumnName = parentTableAlias + "." + childColumnName;
				ZConstant childColumn = new ZConstant(childColumnName, ZConstant.COLUMNNAME);

				String parentColumnName = joinCondition.getParentColumnName();
				parentColumnName = parentColumnName.replaceAll("\"", enclosedCharacter);
				parentColumnName = joinQueryAlias + "." + parentColumnName;
				ZConstant parentColumn = new ZConstant(parentColumnName, ZConstant.COLUMNNAME);
				
				ZExpression joinConditionExpression = new ZExpression("=", childColumn, parentColumn);
				if(onExpression == null) {
					onExpression = joinConditionExpression;
				} else {
					onExpression = new ZExpression("AND", onExpression, joinConditionExpression);
				}
			}
		}
		
		return onExpression;
	}	
	
}	

File: morph-r2rml/src/main/scala/es/upm/fi/dia/oeg/morph/r2rml/model/R2RMLJoinCondition.scala
package es.upm.fi.dia.oeg.morph.r2rml.model

import org.apache.jena.rdf.model.Resource
import es.upm.fi.dia.oeg.morph.base.Constants
import org.slf4j.LoggerFactory

class R2RMLJoinCondition(val childColumnName:String , val parentColumnName:String ) {
  val logger = LoggerFactory.getLogger(this.getClass());

	override def toString() : String = {
		return "JOIN(Parent:" + this.parentColumnName + ",CHILD:" + this.childColumnName + ")";
	}
}

object R2RMLJoinCondition {
  val logger = LoggerFactory.getLogger(this.getClass());
	
	def apply(resource:Resource ) : R2RMLJoinCondition = {
		val childStatement = resource.getProperty(Constants.R2RML_CHILD_PROPERTY);
		if(childStatement == null) {
			val errorMessage = "Missing child column in join condition!";
			logger.error(errorMessage);
			throw new Exception(errorMessage);		  
		}
		val childColumnName = childStatement.getObject().toString();

		val parentStatement = resource.getProperty(Constants.R2RML_PARENT_PROPERTY);
		if(parentStatement == null) {
			val errorMessage = "Missing parent column in join condition!";
			logger.error(errorMessage);
			throw new Exception(errorMessage);		  
		}
		val parentColumnName = parentStatement.getObject().toString();
		
		new R2RMLJoinCondition(childColumnName, parentColumnName); 
	}
  

}
